File: Task 4/Java/app/src/main/java/dev/ogabek/java/manager/DatabaseManager.java
package dev.ogabek.java.manager;

import androidx.annotation.NonNull;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

import dev.ogabek.java.model.Post;

public class DatabaseManager {

    private static final DatabaseReference database = FirebaseDatabase.getInstance().getReference();
    private static final DatabaseReference reference = database.child("posts");

    public static void storePost(Post post, DatabaseHandler handler) {
        String key;
        if (reference.push().getKey() == null) {
            return;
        } else {
            key = reference.push().getKey();
        }
        post.setId(key);
        reference.child(key).setValue(post).addOnCompleteListener(task -> {
            handler.onSuccess();
        }).addOnFailureListener(e -> {
            handler.onError();
        });
    }
    public static void apiLoadPosts(DatabaseHandler handler) {
        ArrayList<Post> posts = new ArrayList<>();
        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                if (snapshot.exists()) {
                    for (DataSnapshot i : snapshot.getChildren()) {
                        Post post = i.getValue(Post.class);
                        if (post != null) {
                            posts.add(post);
                        }
                    }
                    handler.onSuccess(posts);
                } else {
                    handler.onSuccess(new ArrayList<>());
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                handler.onError();
            }
        });
    }

}

File: Task 1/Kotlin/app/src/main/java/dev/ogabek/kotlin/manager/DatabaseManager.kt
package dev.ogabek.kotlin.manager

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import dev.ogabek.kotlin.model.Post

class DatabaseManager {

    companion object {

        private val database = FirebaseDatabase.getInstance().reference
        private val reference = database.child("posts")

        fun storePost(post: Post, handler: DatabaseHandler) {
            val key = reference.push().key ?: return
            post.id = key
            reference.child(key).setValue(post).addOnSuccessListener {
                handler.onSuccess()
            }.addOnFailureListener {
                handler.onError()
            }
        }

        fun apiLoadPosts(handler: DatabaseHandler) {
            reference.addValueEventListener(object : ValueEventListener {
                val posts: ArrayList<Post> = ArrayList()
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (snapshot.exists()) {
                        for (i in snapshot.children) {
                            val post: Post? = i.getValue(Post::class.java)
                            post.let {
                                posts.add(post!!)
                            }
                        }
                        handler.onSuccess(posts = posts)
                    } else {
                        handler.onSuccess(posts = ArrayList())
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                    handler.onError()
                }

            })
        }



    }

}